### PR TITLE
Refactor shader code into separate modules and add some shader tests

### PR DESCRIPTION
**Pull request by philippotto**
TLDR: The goal of this PR is to refactor the shader code into smaller chunks that are easier to understand and to test.

In more detail, this PR:
- implements a lightweight module system for shader code (each module can define code and dependencies which can be resolved)
- refactors the shader code into small modules
- adds a glsl parser which is used in ava tests to validate compiled shader code
- adds headless-gl as a dependency which is used in ava tests to render shader code

In the long-term, I'd like to:
- Get rid of lodash templates (use es6 template literals instead) to get syntax and type checking with flow & co.
- Be able to render an actual bucket (needs actual uniform binding which is too much for this PR)
- Make more use of shader recompilation to minimize shader size. E.g., the mappings code could be excluded from compilation until a mapping is activated.

### To Discuss
- [x] The shader modules live in `*.glsl.js` files. Ideally, I'd like to configure my editor to use glsl syntax for these files, but I'm afraid most editors cannot do that unless the ending is `.glsl`. Defining the files like that would be possible, too, but it's probably a hassle to configure that throughout all build tools (webpack, prettier, linter etc.)
- [x] On circleci, headless-gl needs full compilation which takes approx. 3 minutes. This is way too long in my opinion. We should investigate options to [prebuild](https://github.com/prebuild/prebuild) (?) this. This should be done before merging this PR imo.

### Steps to test:
- open different kind of tracings (with/without segmentation; multiple layers etc.)
- use view, skeleton & volume tracing
- for skeletons: use ortho, flight & oblique mode

### Issues:
- contributes to #2615 

------
- [x] Ready for review

### CODE REVIEW

**Comment by philippotto** [comment]
I forked headless-gl to our organization and compiled the necessary binaries so that the CI doesn't need to re-compile it on every build.

It was a bit tricky to set it up in general, but adding new binaries should be relatively simple now. For 
the record, here's how it's done:
- checkout scalableminds/headless-gl on architecture you want to support (e.g., in docker container by running `docker-compose run dev bash`)
- ensure dependencies are installed:
	- apt-get update
	- apt-get install mesa-utils xvfb libgl1-mesa-dri libglapi-mesa libosmesa6 pkg-config x11proto-xext-dev xserver-xorg-dev libxext-dev libxi-dev

- run node_modules/.bin/prebuild --strip --verbose to compile binaries for current platform
- run node_modules/.bin/prebuild --upload-all to attach the built binaries (exist in prebuilds folder) to the newest release

**Comment by philippotto** [comment]
The PR is ready for review @daniel-wer. Feel free to discuss and question the general approach :)

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 FROM openjdk:8-jdk
 RUN apt-get update \
-    && apt-get -y install postgresql-client \
+    && apt-get -y install postgresql-client pkg-config \
```
Comment (jstriebel) [inline Dockerfile]: isn't this only necessary in the CI?

Comment (philippotto) [reply Dockerfile]: Oops, sorry, I didn't mean to commit that.

File: package.json [hunk 2]
```diff
@@ -100,6 +104,8 @@
     "es6-promise": "^3.0.2",
     "file-saver": "^1.3.3",
     "font-awesome": "^4.5.0",
+    "glsl-parser": "git+https://github.com/anvaka/glslx.git#glsl-parser",
```
Comment (daniel-wer) [inline package.json]: What's the reason for using this fork? :)

File: package.json [hunk 3]
```diff
@@ -100,6 +104,8 @@
...
+    "glslify": "^6.1.1",
```
Comment (daniel-wer) [inline package.json]: I didn't see this being used in the project, is it needed explicitly for glsl-parser?

File: app/assets/javascripts/test/shaders/shader_syntax.spec.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    isRgb: false,
+    planeID: OrthoViews.PLANE_XY,
+    isMappingSupported: true,
+    dataTextureCountPerLayer: 3,
+    resolutions,
+    datasetScale: [1, 1, 1],
+  });
+  console.log("code.length", code.length);
+
+  console.time("check");
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply app/assets/javascripts/test/shaders/shader_syntax.spec.js]: I saw that you included several console.time lines, if we want to keep them, I would suggest to add a prefix to understand the output (e.g.: `console.time("shader-test.check-shader")` or so).

**Comment by daniel-wer** [comment]
From a quick look it would be easy to configure the `.glsl` file ending for eslint (https://eslint.org/docs/user-guide/command-line-interface#--ext), webpack (webpack.config.js - modules.rules.test) and flow (https://github.com/facebook/flow/pull/562), but I didn't find anything for prettier.

**Comment by daniel-wer** [comment]
I found a solution to highlight `.glsl.js` files in sublime using c/c++ syntax highlighting, without affecting other `.js` files:

- Open a `.glsl.js` file
- Activate the Syntax Highlighting you want to use (e.g. View -> Syntax -> C)
- Open the syntax settings (Preferences -> Settings Syntax Specific)
- Add the string `glsl.js` to the extensions list (No leading `.`!)
- Restart sublime, `.glsl.js` files should be opened with the correct syntax highlighting from now on
- :)

File: package.json (hunk 2 above)
Comment (philippotto) [reply package.json]: This fork exposes the actual parser so that we can import it. The original repository doesn't do this unfortunately :/

File: package.json (hunk 3 above)
Comment (philippotto) [reply package.json]: Good catch, this is obsolete.
